### ISSUE TITLE
Shaping output incorrect

### ISSUE THREAD

**Issue by walpoletim-bjss** [xionluhnis/knitsketching#2]
We are attempting to shape the front of a v-neck jumper, however the output is not as expected..

**Input** 
<img width="792" alt="Screenshot 2021-08-31 at 17 49 26" src="https://user-images.githubusercontent.com/10545041/131544211-6a3ed771-676e-4373-9b6b-52efda47c751.png">

**Output** (Using the knitout visualiser, including waste bindon for kniterate machine in purple)
<img width="821" alt="Screenshot 2021-08-31 at 17 48 27" src="https://user-images.githubusercontent.com/10545041/131544345-6e937c99-3ee5-421e-94e4-03fb4ba89659.png">

**Input file to KnitSketching**
[JumperShape.json.txt](https://github.com/xionluhnis/knitsketching/files/7085556/JumperShape.json.txt)

For reference, we are successfully knitting some of the jacquard samples on a kniterate machine, with the kniterate convertor below h
![IMG_5307](https://user-images.githubusercontent.com/10545041/131545482-3c387021-0b7f-401f-bc5d-5a35a2fad2d3.jpg)
https://github.com/textiles-lab/knitout-backend-kniterate

Knitted Output
![IMG_5308](https://user-images.githubusercontent.com/10545041/131548784-58b39311-1393-493e-ba96-34e1f70a6695.jpg)

**Comment by xionluhnis** [comment]
This works almost as expected, modulo the unexpected lack of short-rows (though I hope this can be fixed by tuning some parameters).

The main problem is the time isoline constraints at the sketch top boundaries: they all form a single time constraint, and this is probably not what you want for a V-neck structure.

There are two notable ways to achieve that "v-neck":
1. By having the bottom of the V split the region (1-flat to 2-flats, or tubular to c-shaped flat);
2. By using short-rows to make the shape rotate towards the front (using suspended stitches on the left and right sides).

## Option 1: v-neck with region split

To achieve (1), you must change the constraints so that the neck end is not an isoline constraint.
Getting a single course that covers the end edge is unfortunately something related to scheduling (not the shape itself), and the scheduler wouldn't support it right now. But at least you should be able to get the proper v-neck shape (modulo the imperfect edge finish).
To get a perpendicular flow, but not a time constraint, you can use the source/sink constraint types (they are not in the paper, but it's basically similar to a direction constraint, except that the direction is orthogonal to the constraint curve).
Though, if you just delete the neck constraints, this already somewhat works:

[JumperShape-noNeckConstr.json.txt](https://github.com/xionluhnis/knitsketching/files/7086016/JumperShape-noNeckConstr.json.txt)
<img width="729" alt="JumperShape-noNeckConstr" src="https://user-images.githubusercontent.com/1904312/131554733-abc0e6d5-65c6-452d-b92c-8f1a68cf8768.png">

However, if you create the stitch graph, it won't schedule properly.
This is an issue due to the 1-flat to 2-flat splitting.
The scheduler I implemented is incomplete as it does not support "splitting".
The code could be extended to support it, but you typically don't want to split, especially if you can do it in the opposite direction and "merge" instead.

There's fortunately a simple button for that: below the program button, the "Graph" box has "inverse" which reverses the computed time function. Then it schedules properly and you get this:
<img width="826" alt="JumperShape-noNeckConstr2" src="https://user-images.githubusercontent.com/1904312/131554737-2d475e2b-60d9-4cee-8bc0-41b79c98866b.png">
<img width="185" alt="JumperShape-noNeckConstr3" src="https://user-images.githubusercontent.com/1904312/131554746-254eb4f4-0e20-403c-81fb-d83ba45985c8.png">

Shima's templates would allow for splitting but use parallel yarn carrier sections, and the current schedulers don't do that either.
The new scheduling mechanism of the work of Nader would likely enable such more "ideal" scheduling.
Though there's no code for it yet, and it's a very different data structure, so it's unclear how much work would be needed to support that in the current system.
https://gnader.github.io/publications/2021-knitkit.html

## Option 2: v-neck with short-rows

To achieve (2), you need to tweak the constraints to avoid the short-rows being diffused (and currently *cancelled* as a result).
This can be achieved with two ways: (i) by introducing a time isoline constraint in the body section that compacts the upper region so that short-rows appear un-diffused; and (ii) by explicitly modeling the time dilation in that region with curvature constraints (not part of the main paper, but described in the supplementary document).
This is not the conventional way of doing a v-neck, but it should be possible, although it's unfortunately one case of high-curvature that is unstable, as described in the paper / supplement.
Without the curvature constraints (i.e., just using the intermediate isoline constraint) the system will still converge to a diffuse solution that fails at getting proper short-rows.

Using curvature constraints (e.g. k=0.3) should induce the necessary short-rows and stabilize the time function computation.
I attempted to test that in your example, but it seems to trigger some error when computing the regions. The time function seems good, but the regions crash the system.
This is likely a bug.

[JumperShape-curvature.json.txt](https://github.com/xionluhnis/knitsketching/files/7086095/JumperShape-curvature.json.txt)
<img width="701" alt="JumperShape-curvature" src="https://user-images.githubusercontent.com/1904312/131557623-8b5c5710-6b6d-408d-89bf-2b173ce47753.png">

**Comment by walpoletim-bjss** [comment]
Amazing - Thank you so much for the detailed response - Really useful.

Will read throughly tomorrow and run through the machine. 

BTW - Your software is amazing :)

**Comment by walpoletim-bjss** [comment]
Hi
By using your suggestions, we have indeed managed to produce a knitout that knits each side of the neck independently.  
This however, on a Kniterate machine would not knit successfully as the take down rollers operate across the whole bed, and so it is not possible to knit each side independently.

As such we would require parallel knitting of each side with different feeders as you have suggested.

Would be great to see this being implemented at some point:):)

**Comment by xionluhnis** [comment]
Good to know.
But unfortunate for kniterate users right now.

On the Shima I used, I did not need the rollers (or waste fabric for them) thanks to the sinker system.
Though it would be good to support it even for the Shima target since you can potentially knit faster by knitting all sections in parallel (and it'd be necessary for complex split/merge structures).
